fix: fix backreferencing issue for digits greater than 10

jeswr opened the pull request:
Closes #38 .

This solves the initial issue discussed in #38 , however, it does not yet resolve this [follow up](https://github.com/fent/ret.js/issues/38#issuecomment-1006460390) observation.

In fact this introduces a bug for patterns like `(a) - \11` as `\11` *should* be treated as a value since there is only one capturing group. I will leave this as a draft until this is resolved.

fent commented:
> In fact this introduces a bug for patterns like `(a) - \11` as `\11` _should_ be treated as a value since there is only one capturing group. I will leave this as a draft until this is resolved.

i checked if you can reference a group before it shows up in the regex

![image](https://user-images.githubusercontent.com/933490/148372322-789a011f-21e9-48a2-96d9-c33f14779c11.png)

so this literal value would have to be inserted post parsing since it needs to know how many groups exist

jeswr commented:
> so this literal value would have to be inserted post parsing since it needs to know how many groups exist

This is a very good point/find - I have solved everything except this as of https://github.com/fent/ret.js/pull/39/commits/411b848c70c90b704ee0b8c940b5af222b1fc45c but cannot think of an elegant way to do this post processing right now (I accidentally staged a broken attempt at doing this in my most recent lint:fix commit but you can ignore this).

I don't have time to work on resolving this for the next couple of days - so if you want to take over or point me in a direction that you are thinking it would be appreciated :)

jeswr commented:
Converting back to draft as I made another incorrect assumption about the behavior of capture groups - they are not scoped. I.e.you can reference outer capture groups from inside another one

![image](https://user-images.githubusercontent.com/63333554/148478076-20de63d7-a77f-4745-8fc8-6a4b56962aae.png)

fent commented on lib/tokenizer.ts:
```diff
@@ -17,10 +17,12 @@ export const tokenizer = (regexpStr: string): Root => {
     throw new SyntaxError(
-      `Invalid regular expression: /${
-        regexpStr
+      `Invalid regular expression: /${regexpStr
```
this PR has many small stylistic changes like this. were these auto formatted? i ran the `lint:fix` script but it didn't generate these changes. even so, let's try to keep this PR's changes related to only the PR's purpose. otherwise, it is harder to read

jeswr replied:
This was probably the prettier extension I have in vscode. I have now reverted these formatting changes :)

fent commented on lib/tokenizer.ts:
```diff
@@ -154,6 +168,8 @@ export const tokenizer = (regexpStr: string): Root => {
 
           group.remember = false;
+        } else {
+          groupCount += 1;
```
oh i hadn't realized they can't be referenced if they're special group types. cool

fent commented:
thanks for the fix! learned some new things about javascript regex :o

fent reviewed (changes requested):
wait! i was just looking at https://github.com/fent/ret.js/issues/31

the escaped code points only match octal numbers, so

    /\18/ // Match code point 1, followed by "8"

jeswr commented:
> wait! i was just looking at #31
> 
> the escaped code points only match octal numbers, so
> 
> ```
> /\18/ // Match code point 1, followed by "8"
> ```

Unless I'm misinterpreting, this is not correct (at least in firefox)

![image](https://user-images.githubusercontent.com/63333554/148669807-4fffcf7a-1e3a-4794-b349-e886cf893979.png)

fent commented:
> Ahh so this should be tokenized as
> 
> ```ts
> {
> type: types.CHAR,
> value: 1
> },
> {
> type: types.CHAR,
> value: '8'.charCodeAt(0)
> }
> ```

right, makes the `updateReferences` function a bit more complicated. the detection and separation part not so much, but inserting an additional token next to that one is

fent commented on lib/tokenizer.ts:
```diff
@@ -295,5 +313,49 @@ export const tokenizer = (regexpStr: string): Root => {
+      // If the number is not octal then we need to create multiple tokens
+      // https://github.com/fent/ret.js/pull/39#issuecomment-1008229226
+      if (!/^[0-7]+$/.test(valueString)) {
+        const val = parseInt(valueString[0], 10);
```
i don't think the separation occurs only at the 0th position

![image](https://user-images.githubusercontent.com/933490/148674079-202788ca-a096-4626-942e-ed617a724082.png)

jeswr replied:
I *think* this has been addressed properly now

fent commented on lib/tokenizer.ts:
```diff
@@ -295,5 +313,49 @@ export const tokenizer = (regexpStr: string): Root => {
+        if (valueString.length > 1) {
+          const tail = elem.stack.splice(elem.index + 1);
+
+          // eslint-disable-next-line max-depth
```
i think we can remove this rule instead. too inconvenient sometimes. what do you think?

fent commented on test/main-test.js:
```diff
@@ -465,6 +465,310 @@ vows.describe('Regexp Tokenizer')
+          });
+        },
+      },
+      '\\9': {
```
i would add an additional test for `'\\90'`, should tokenize as `[char('9'), char('0')`

fent reviewed (changes requested):
finally looking at this. added a couple of comments on tests. otherwise, the logic looks good!

jeswr commented:
> finally looking at this. added a couple of comments on tests. otherwise, the logic looks good!

Added those changes - should be ready to merge now

fent commented:
thanks again!
